File: TugasModul4/src/kehadiran_subscriber.cpp
#include "ros/ros.h"
#include "package1/pesan.h"
#include "ascii-art-main/Ascii.h"

using namespace ascii;

std::string ros_ascii_info(std::string text) {
    // Membuat ASCII art
    Ascii font = Ascii(boomer);
    font.print(text);
    std::string letters;
    for (int i = 0; i < font.getletters().size(); i++)
    {
        for (int j = 0; j < font.getletters()[i].size(); j++)
        {
            letters += font.getletters()[i][j];
        }
        letters += "\n";
    }
    return letters;
}

void callbackHandler(const package1::pesan::ConstPtr &msg) {
    // Mencetak pesan yang diterima
    ROS_INFO("\n%s\nKEHADIRAN TERDENGAR:\nusia: %ld\ntinggi: %lf\nnama: %s\nkehadiran: %s\n ",
        ros_ascii_info(msg->kehadiran ? "HADIR" : "ABSEN").c_str(),
        msg->usia, 
        msg->tinggi, 
        msg->nama.c_str(), 
        msg->kehadiran ? "hadir" : "tidak hadir"
        );
}

int main(int argc, char ** argv){
    // Inisialisasi Node
    ros::init(argc,argv,"kehadiran_subscriber");

    // Inisialisasi Handler Node
    ros::NodeHandle nh;

    // Insialisasi Subscriber dari Handler
    ros::Subscriber data_subscriber = nh.subscribe("kehadiran", 1000, callbackHandler);

    ros::spin();

    return 0;
}

File: TugasModul4/script/kehadiran_subscriber.py
#!/usr/bin/env python3
import rospy
from package1.msg import pesan

def showHadir():
    text = (
"""
_   _     ___     ______   _____   ______
| | | |   / _ \   |  _  \  |_   _|  | ___ \ 
| |_| |  / /_\ \  | | | |    | |    | |_/ /
|  _  |  |  _  |  | | | |    | |    |    /
| | | |  | | | |  | |/ /    _| |_   | |\ \ 
\_| |_/  \_| |_/  |___/     \___/   \_| \_|
"""
    )
    return text

def showAbsen():
    text = (
"""
  ___    ______    _____    _____    _   _
 / _ \   | ___ \  /  ___|  |  ___|  | \ | |
/ /_\ \  | |_/ /  \ `--.   | |__    |  \| |
|  _  |  | ___ \   `--. \  |  __|   | . ` |
| | | |  | |_/ /  /\__/ /  | |___   | |\  |
\_| |_/  \____/   \____/   \____/   \_| \_/
"""
    )
    return text

def callback(data):
    # Mencetak pesan yang diterima
    if data.kehadiran:
        head = showHadir()
    else:
        head = showAbsen()
    
    rospy.loginfo("\n%s\nTERDENGAR usia: %ld tinggi: %lf nama: %s kehadiran: %s" % 
                  (head, 
                   data.usia, 
                   data.tinggi, 
                   data.nama, 
                   "Hadir" if data.kehadiran else "Tidak hadir"))
    
def listener():
    # Inisialisasi Node
    rospy.init_node("kehadiran_subscriber",anonymous=True)

    # Inisialisasi Subscriber
    rospy.Subscriber("kehadiran", pesan, callback)

    rospy.spin()

if __name__ == "__main__":
    try:
        listener()
    except rospy.ROSInterruptException:
        pass